Libs/MRML/Core/vtkMRMLTableStorageNode.cxx

```diff
@@ -340,106 +340,273 @@ bool vtkMRMLTableStorageNode::ReadTable(std::string filename, vtkMRMLTableNode*
+  if (schema != nullptr)
+    {
+    schemaColumnNameArray = vtkStringArray::SafeDownCast(schema->GetColumnByName("columnName"));
+    schemaComponentNamesArray = vtkStringArray::SafeDownCast(schema->GetColumnByName("componentNames"));
+    }
+
+  // Populate the output table column details.
+  // If the schema exists, read the contents and determine column data type/component names/component arrays
+  if (schema != nullptr &&
+      schemaColumnNameArray != nullptr &&
+      schemaComponentNamesArray != nullptr)
     {
-    vtkStringArray* column = vtkStringArray::SafeDownCast(rawTable->GetColumn(col));
-    if (column == nullptr)
+    for (int schemaRowIndex = 0; schemaRowIndex < schema->GetNumberOfRows(); ++schemaRowIndex)
       {
-      // invalid column
-      continue;
+      ColumnInfo columnInfo;
+      columnInfo.ColumnName = schemaColumnNameArray->GetValue(schemaRowIndex);
+      columnInfo.DataType = tableNode->GetColumnValueTypeFromSchema(columnInfo.ColumnName);
+
+      std::vector<vtkAbstractArray*> componentArrays;
+      std::string componentNamesStr = schemaComponentNamesArray->GetValue(schemaRowIndex);
+      if (componentNamesStr.empty())
+        {
+        vtkAbstractArray* rawColumn = rawTable->GetColumnByName(columnInfo.ColumnName.c_str());
+        componentArrays.push_back(rawColumn);
+        }
+      else
+        {
+        std::stringstream ss(componentNamesStr);
+        std::string componentName;
+        while (std::getline(ss, componentName, '|'))
+          {
+          std::string componentColumnName = columnInfo.ColumnName + "_" + componentName;
+          vtkAbstractArray* rawColumn = rawTable->GetColumnByName(componentColumnName.c_str());
+          componentArrays.push_back(rawColumn);
+          columnInfo.ComponentNames.push_back(componentName);
+          }
+        }
+      columnInfo.RawComponentArrays = componentArrays;
+      columnDetails.push_back(columnInfo);
       }
-    if (!column->GetName())
+    }
+  else
+    {
+    for (int col = 0; col < rawTable->GetNumberOfColumns(); ++col)
       {
-      vtkWarningMacro("vtkMRMLTableStorageNode::ReadTable: empty column name in file: " << filename << ", skipping column");
-      continue;
+      ColumnInfo columnInfo;
+      vtkStringArray* column = vtkStringArray::SafeDownCast(rawTable->GetColumn(col));
+      if (column == nullptr)
+        {
+        // invalid column
+        continue;
+        }
+      if (!column->GetName())
+        {
+        vtkWarningMacro("vtkMRMLTableStorageNode::ReadTable: empty column name in file: " << filename << ", skipping column");
+        continue;
+        }
+      columnInfo.ColumnName = column->GetName();
+      columnInfo.DataType = tableNode->GetColumnValueTypeFromSchema(columnInfo.ColumnName);
+      columnInfo.RawComponentArrays.push_back(column);
+      columnDetails.push_back(columnInfo);
       }
+    }
+
+  vtkSmartPointer<vtkTable> table = vtkSmartPointer<vtkTable>::New();
```

Comment:
Add 1-2 summary sentences about what is going to happen here. Since the method is a bit too long, you may even consider to split it into a few methods. This is a potential split point position.